curve-amm/src/lib.rs

```diff
@@ -0,0 +1,989 @@
+        let y = self.get_y(i, j, x, &xp, ann).ok_or(CurveAmmError::Math)?;
+        let dy = xp[j]
+            .checked_sub(&y)
+            .and_then(|v| v.checked_sub(&prec))
+            .ok_or(CurveAmmError::Math)?;
+
+        let pool = self.get_pool(&pool_id)?;
+        let lock = MUTEX.lock().await;
+        let reply: Result<Event, ContractError> = msg::send_and_wait_for_reply(
+            pool.assets[i],
+            &Action::BalanceOf(*who),
+            100_000_000_000,
+            0,
+        )
+        .await;
+        gstd::mem::drop(lock);
+        let reply = reply.map_err(CurveAmmError::BalaceOfFailed)?;
+        let balance = match reply {
+            Event::Balance(bal) => bal,
+            _ => {
+                return Err(CurveAmmError::DecodeError);
+            }
+        };
+        let balance: FixedU128 = Self::u128_to_fixed(balance)?;
+        if balance < dx {
+            return Err(CurveAmmError::InsufficientFunds);
+        }
+        let lock = MUTEX.lock().await;
+        let reply: Result<Event, ContractError> = msg::send_and_wait_for_reply(
+            pool.assets[j],
+            &Action::BalanceOf(exec::program_id()),
+            100_000_000_000,
+            0,
+        )
+        .await;
+        gstd::mem::drop(lock);
+        let reply = reply.map_err(CurveAmmError::BalaceOfFailed)?;
+        let balance = match reply {
+            Event::Balance(bal) => bal,
+            _ => {
+                return Err(CurveAmmError::DecodeError);
+            }
+        };
+        let balance: FixedU128 = Self::u128_to_fixed(balance)?;
+        let amount: u128 = Self::fixed_to_u128(&dx);
+        if balance < dy {
+            return Err(CurveAmmError::InsufficientFunds);
+        }
+        let lock = MUTEX.lock().await;
+        let reply: Result<Event, ContractError> = msg::send_and_wait_for_reply(
+            pool.assets[i],
+            &Action::TransferFrom(TransferFromInput {
+                owner: *who,
+                to: exec::program_id(),
+                amount,
+            }),
+            100_000_000_000,
+            0,
+        )
+        .await;
+        gstd::mem::drop(lock);
+        let _reply = reply.map_err(CurveAmmError::TransferFromFailed)?;
+        let amount: u128 = Self::fixed_to_u128(&dy);
+        let lock = MUTEX.lock().await;
+        let reply: Result<Event, ContractError> = msg::send_and_wait_for_reply(
+            pool.assets[j],
+            &Action::Transfer(TransferInput { to: *who, amount }),
+            100_000_000_000,
+            0,
+        )
+        .await;
+        gstd::mem::drop(lock);
+        let _reply = reply.map_err(CurveAmmError::TransferFailed)?;
+        let exchange_reply = CurveAmmExchangeReply {
+            who: *who,
+            pool_id,
+            i: i_u,
+            j: j_u,
+            dy_amount: amount,
+        };
+
+        msg::reply(CurveAmmReply::Exchange(exchange_reply), 0, 0);
+        Ok(())
+    }
+}
```

Comment:
All this code related to AMM could be moved to the standalone model, `curve_amm` for example, since it implements all business AMM-logic.

Then lib.rs would contain just a couple of globals, `init` and `main`.

I'm not insisting though.

Reply:
I agree that all math curve based AMM like Uniswap and StableSwap has almost similar interface like add/remove liquidity, exchange(), buy/sell etc but their math is different. So for now I am not separating it. Although if we want to have Uniswap-V2 implementation I can refactor to make it more like library.